lib/e2ee/e2ee.h

```diff
@@ -0,0 +1,132 @@
+inline const auto RotationPeriodMsKeyL = "rotation_period_ms"_ls;
+inline const auto RotationPeriodMsgsKeyL = "rotation_period_msgs"_ls;
+
+inline const auto AlgorithmKey = QStringLiteral("algorithm");
+inline const auto RotationPeriodMsKey = QStringLiteral("rotation_period_ms");
+inline const auto RotationPeriodMsgsKey =
+    QStringLiteral("rotation_period_msgs");
+
+inline const auto Ed25519Key = QStringLiteral("ed25519");
+inline const auto Curve25519Key = QStringLiteral("curve25519");
+inline const auto SignedCurve25519Key = QStringLiteral("signed_curve25519");
+inline const auto OlmV1Curve25519AesSha2AlgoKey =
+    QStringLiteral("m.olm.v1.curve25519-aes-sha2");
+inline const auto MegolmV1AesSha2AlgoKey =
+    QStringLiteral("m.megolm.v1.aes-sha2");
+inline const QStringList SupportedAlgorithms = { OlmV1Curve25519AesSha2AlgoKey,
+                                                 MegolmV1AesSha2AlgoKey };
+struct Unencrypted {};
+struct Encrypted {
+    QByteArray key;
+};
+
+using PicklingMode = std::variant<Unencrypted, Encrypted>;
+
+class QOlmSession;
+using QOlmSessionPtr = std::unique_ptr<QOlmSession>;
+
+class QOlmInboundGroupSession;
+using QOlmInboundGroupSessionPtr = std::unique_ptr<QOlmInboundGroupSession>;
+
+template<class... Ts> struct overloaded : Ts... { using Ts::operator()...; };
+template<class... Ts> overloaded(Ts...) -> overloaded<Ts...>;
+
+struct IdentityKeys
+{
+    QByteArray curve25519;
+    QByteArray ed25519;
+};
+
+//! Struct representing the one-time keys.
+struct OneTimeKeys
+{
+    QMap<QString, QMap<QString, QString>> keys;
+
+    //! Get the HashMap containing the curve25519 one-time keys.
+    QMap<QString, QString> curve25519() const;
+
+    //! Get a reference to the hashmap corresponding to given key type.
+    std::optional<QMap<QString, QString>> get(QString keyType) const;
+};
+
+//! Struct representing the signed one-time keys.
+class SignedOneTimeKey
+{
+public:
+    SignedOneTimeKey() = default;
+    SignedOneTimeKey(const SignedOneTimeKey &) = default;
+    SignedOneTimeKey &operator=(const SignedOneTimeKey &) = default;
+    //! Required. The unpadded Base64-encoded 32-byte Curve25519 public key.
+    QString key;
+
+    //! Required. Signatures of the key object.
+    //! The signature is calculated using the process described at Signing JSON.
+    QHash<QString, QHash<QString, QString>> signatures;
+};
+
+
+template <>
+struct JsonObjectConverter<SignedOneTimeKey> {
+    static void fillFrom(const QJsonObject& jo,
+                         SignedOneTimeKey& result)
+    {
+        fromJson(jo.value("key"_ls), result.key);
+        fromJson(jo.value("signatures"_ls), result.signatures);
+    }
+
+    static void dumpTo(QJsonObject &jo, const SignedOneTimeKey &result)
+    {
+        addParam<>(jo, QStringLiteral("key"), result.key);
+        addParam<>(jo, QStringLiteral("signatures"), result.signatures);
+    }
+};
+
+bool operator==(const IdentityKeys& lhs, const IdentityKeys& rhs);
+
```

Comment:
And this seems to lack the definition - however, apparently is unused either? I don't see any linkers I tried complaining.

Reply:
Sorry, found it in `qolmaccount.cpp`. I still don't see it being used anywhere, so will probably try to go without one, for now.